tests/test_sync_bookmark.py

```diff
@@ -0,0 +1,320 @@
+    return result
+
+
+@pytest.fixture()
+def generate_data(dataframe_factory, sync_summary_schema):
+    def _generate_data(snippets):
+        return (
+            dataframe_factory.create_dataframe(
+                map(build_sync_summary_snippet, snippets),
+                sync_summary_sample,
+                sync_summary_schema
+            )
+        )
+    return _generate_data
+
+
+@pytest.fixture()
+def test_transform(spark, generate_data, monkeypatch):
+    def _test_transform(snippets):
+        def mock_parquet(cls, path):
+            return generate_data(snippets)
+        monkeypatch.setattr(
+            "pyspark.sql.DataFrameReader.parquet",
+            mock_parquet
+        )
+        sbv.extract(spark, None, "20171001")
+        sbv.transform(spark)
+        return spark.table("bmk_validation_problems"), \
+            spark.table("bmk_total_per_day")
+    return _test_transform
+
+
+def test_failures_filtered(test_transform):
+    df, _ = test_transform([
+        {'failure_reason': None},
+        {'failure_reason': {"name": "failure!"}},
+        {'failure_reason': {"name": "failure2!"}},
+    ])
+
+    assert df.count() == 1
+
+
+def test_validation_problems(test_transform):
+    df, _ = test_transform([
+        # failed, also not a bookmark
+        {
+            'failure_reason': {'name': 'some failure'},
+            'engines': [{
+                'name': 'not bookmarks',
+                'validation': {'problems': None}
+            }]
+        },
+        # not a bookmark
+        {
+            'failure_reason': None,
+            'engines': [{
+                'name': 'not bookmarks',
+                'validation': {'problems': None}
+            }]
+        },
+        # does not contain a problem
+        {
+            'failure_reason': None,
+            'engines': [{
+                'name': 'bookmarks',
+                'validation': {'problems': None}
+            }]
+        },
+        # single engine, single problem, no longer explicitly setting values
+        {
+            'engines': [{
+                'validation': {
+                    'problems': [{'name': '1', 'count': 1}]
+                }
+            }]
+        },
+        # two engines, with one and two problems each
+        {
+            'engines': [
+                # no problems
+                {'name': 'not bookmarks'},
+                # a single problem
+                {
+                    'validation': {
+                        'problems': [{'name': '2', 'count': 10}]
+                    }
+                },
+                {
+                    'validation': {
+                        'problems': [
+                            {'name': '3', 'count': 100},
+                            {'name': '4', 'count': 1000},
+                        ]
+                    }
+                }
+            ]
+        },
+    ])
+
+    assert df.count() == 4
+
+    # one hot encode the cases
```

Comment:
nit: I don't understand this comment

Reply:
You can tell which case failed based on the digit. Case 1 corresponds to digit 1 (0001), case 3 corresponds to digit 3 (0100). I still had to drop into pdb to debug the cases though.